ui/src/table.tsx

```diff
@@ -735,7 +735,7 @@ export const
 
         setGroups(groups => {
           if (groups) {
-            setFilteredItems(filteredItems => groups?.reduce((acc, group) =>
+            setFilteredItems(filteredItems => [...groups].sort((group1, group2) => group1.startIndex - group2.startIndex).reduce((acc, group) =>
```

Comment:
Why does this work?

Reply:
The original implementation supposes that value of `groups` is similar to this one:
```ts
const groups = [
  { 
  key: "groupC", 
  name: "groupC", 
  startIndex: 0, 
  count: 3, 
  isCollapsed: false
  },
  { 
  key: "groupA", 
  name: "groupA", 
  startIndex: 3, 
  count: 2, 
  isCollapsed: false
  },
  { 
  key: "groupB", 
  name: "groupB", 
  startIndex: 5, 
  count: 4, 
  isCollapsed: false
  }
]
```
However, this is not true since `groups` are sorted alphabetically [on line 562](https://github.com/h2oai/wave/blob/56a9ce8f2ef3a48bc3a956aa178e16de1aaf3acc/ui/src/table.tsx#L562) in `makeGroups`. Therefore the real value of `groups` is in our case this:
```ts
const groups = [
  { 
  key: "groupA", 
  name: "groupA", 
  startIndex: 3, 
  count: 2, 
  isCollapsed: false
  },
  { 
  key: "groupB", 
  name: "groupB", 
  startIndex: 5, 
  count: 4, 
  isCollapsed: false
  },
  { 
  key: "groupC", 
  name: "groupC", 
  startIndex: 0, 
  count: 3, 
  isCollapsed: false
  }
]
```

There is no problem with viewing groups, Fluent takes it whether the value is as in the first example or the second one, but when it comes to sorting, here lies a problem.
Let's have a look at our original group sorting implementation:
```ts
if (groups) {
 setFilteredItems(filteredItems => groups.reduce((acc, group) =>
 [...acc, ...filteredItems.slice(group.startIndex, acc.length + group.count).sort(sortingF(column, sortAsc))],
 [] as any[]) || [])
}
```
We can start with the first iteration of our reducer function:

1. it takes the first group from our `groups` const
```
  { 
  key: "groupA", 
  name: "groupA", 
  startIndex: 3, 
  count: 2, 
  isCollapsed: false
  }
```
2. It slices our `filteredItems` from index 3 to 5
3. It sorts the items
4. And here is the main problem - it pushes it **to the beginning** of the reducer's accumulator (not back to index 3-5!!)

When all iterations are finished, `filteredItems` where subsets are not just sorted, but also rearranged, are set with `setFilteredItems` and the same `groups` are applied to it. Now our `groups` do not match updated `filteredItems`.

The most simple solution here is to sort the `groups` by its `startIndex` before reducing `filteredItems` without modifying the original `groups` value:
```ts
if (groups) {
 setFilteredItems(filteredItems => [...groups].sort((group1, group2) => group1.startIndex 
 group2.startIndex).reduce((acc, group) =>
 [...acc, ...filteredItems.slice(group.startIndex, acc.length + group.count).sort(sortingF(column, sortAsc))],
 [] as any[]) || [])
}
```

Another solution would be to push the sorted items in reducer to the right accumulator's indexes, but this solution would be more costy and complicated.

The reason why the original implementation was working in most of the cases is that usually we have just 2 groups created after groupby in our .py examples and there was a high chance these groups were in correct order.

Reply:
Makes sense, please add a brief comment with explanation for future reference.